backend/tests/get_koji_tasks.py

```diff
@@ -0,0 +1,116 @@
+
+import koji
+
+from src.fetcher import KojiProvider
+
+
+class _KojiFailedBuildsIterator:
+    def __init__(
+        self,
+        koji_client,
+        start_time=None,
+        end_time=None,
+        days_back_limit=30,
+        shared_list=None,
+    ):
+        self._koji_client = koji_client
+        self._curr_index = 0
+        self._days_back_limit = days_back_limit
+
+        self._start_time = start_time
+        if self._start_time is None:
+            self._start_time = (datetime.now() - timedelta(days=1)).timestamp()
+
+        self._end_time = end_time
+        if self._end_time is None:
+            self._end_time = datetime.now().timestamp()
+
+        if shared_list is None:
+            self._shared_list = []
+        else:
+            self._shared_list = shared_list
+            self._sort_shared_list()
+            if self._shared_list:
+                self._start_time = self._minus_days(
+                    self._shared_list[0]["creation_ts"], 1
+                )
+                self._end_time = self._shared_list[-1]["creation_ts"]
+
+    def _sort_shared_list(self):
+        self._shared_list.sort(key=lambda build: build["creation_ts"])
+
+    @staticmethod
+    def _minus_days(curr_time, days):
+        curr_dt = datetime.fromtimestamp(curr_time)
+        return (curr_dt - timedelta(days=days)).timestamp()
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self._days_back_limit == 0:
+            raise StopIteration
+
+        # this is really a crazy number and it shouldn't be possible to hit it
+        if self._curr_index > 10000:
+            raise StopIteration
+
+        if self._curr_index >= len(self._shared_list):
+            self._shared_list.extend(
+                self._koji_client.listBuilds(
+                    completeAfter=self._start_time,
+                    completeBefore=self._end_time,
+                    state=koji.BUILD_STATES["FAILED"],
+                    type="rpm",
+                )
+            )
+            self._sort_shared_list()
+            self._days_back_limit -= 1
+            self._start_time = self._minus_days(self._start_time, 1)
+            self._end_time = self._shared_list[0]["creation_ts"]
+
+        build = self._shared_list[self._curr_index]
+        self._curr_index += 1
+        return build
+
+
+class FailedKojiFetcherTurbo3000:
+    def __init__(self) -> None:
+        self.client = koji.ClientSession(f"{KojiProvider.koji_url}/kojihub")
+        self._shared_failed_builds: list[dict] = []
+
+    def srpm_build_task(self):
+        for build in _KojiFailedBuildsIterator(
+            self.client, shared_list=self._shared_failed_builds
+        ):
+            task_id = build["task_id"]
+            task_descendants = self.client.getTaskDescendents(task_id)[str(task_id)]
+            for desc in task_descendants:
+                if desc["method"] == "buildSRPMFromSCM":
+                    return desc["id"], desc["arch"]
```

Comment:
This seems like it will never end :/ (10 min of searching failed srpm and still nothing...)

Reply:
I'd mock the interaction with koji because it can indeed be very flaky and unreliable

my suggestion is to pick a build of choice and just mock the .getTaskDescendents, getBuild, getTask methods